### PR TITLE
Implement simple copy/move ops for Quarks & kubecf

### PR DESCRIPTION
**Pull request by jandubois**
The copy implementation has been rebased from cppforlife/go-patch#16.

The move op is just a copy of the copy op, and chains to a trailing remove op.

Both copy & move have been renamed to qcopy and qmove to avoid conflicts if the upstream project ever implements copy/move ops with different semantics.

This PR is in support of cloudfoundry-incubator/cf-operator#900.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module github.com/SUSE/go-patch
```
Comment (mook-as) [inline go.mod:1]: If we called this `github.com/cppforlife/go-patch` we wouldn't need to change all the imports, right? Consumers can just use a `replace github.com/cppforlife/go-patch => github.com/SUSE/go-patch` in their go.mod or something?

… Now that I'm typing all this out I'm not so sure :D

Comment (f0rmiga) [reply go.mod:1]: I had a similar suggestion for it as well. My suggestion was to clone the upstream and add the SUSE fork as a git remote. This way we don't even need to add go modules support to this PR (which I don't like very much to be done in this same PR).

Comment (jandubois) [reply go.mod:1]: From https://github.com/golang/go/issues/30354:

>`replace` is intended to support local development, not permanent forks: if you need to permanently fork a dependency, your fork must have its own, unique import path.

So I don't think doing `replace` gymnastics is the right way in this case.

File: patch/copy_op.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package patch
+
+type QCopyOp struct {
```
Comment (f0rmiga) [inline patch/copy_op.go:3]: Missing comments here and below (this and other files) for the exported stuff.

File: patch/move_op.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+	From Pointer
+}
+
+func (op QMoveOp) Apply(doc interface{}) (interface{}, error) {
+	value, err := FindOp{Path: op.From}.Apply(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	doc, err = ReplaceOp{Path: op.Path, Value: value}.Apply(doc)
```
Comment (f0rmiga) [inline patch/move_op.go:14]: Reassigning to a parameter label is not nice. Could you rename the `doc` label to something else, please?

**Review by f0rmiga — CHANGES_REQUESTED** [review]

File: patch/move_op.go (hunk 3 above)
Comment (jandubois) [reply patch/move_op.go:14]: I could, but I intentionally did this to mirror the pattern used in the main Apply code: https://github.com/cppforlife/go-patch/blob/master/patch/ops.go#L21

Each `Apply()` call returns the `doc`, which is then used as the input for the next `Apply()` call until you hit the end.

So I could change it, but I think in this case it makes more sense to re-use it.

File: patch/copy_op.go (hunk 2 above)
Comment (jandubois) [reply patch/copy_op.go:3]: None of the upstream code uses comments on these structs. I could put one in, if it makes you happy, but what should it say: "QCopyOp holds the parameters for a copy operation"?

Isn't that just useless clutter?

Comment (f0rmiga) [reply patch/copy_op.go:3]: Ugh... this is not what the Go convention tells... I'm fine with leaving them out for the sake of consistency with the rest of the repo.

File: patch/move_op.go (hunk 3 above)
Comment (f0rmiga) [reply patch/move_op.go:14]: Yeah, +1 on consistency. Who knows in the future we can just improve the whole codebase for better practices.

**Review by f0rmiga — APPROVED** [review]

**Comment by jandubois** [comment]
Tagged the merge commit with `v0.3.0` for reference from https://github.com/cloudfoundry-incubator/cf-operator/pull/913

**Pull request merged**